File: next-app/src/lib/services/clients.ts
import axios from "axios";
import {createApiWithAuth} from '@/lib/config/axiosConfig'

interface ApiResponse {
    'status': string,
    'message'?: string,
    'clients'?:any[]
}


export const getClients = async() =>{
  try{
    const api = await createApiWithAuth()
    const response = await api.post('/clients/get-all')

    switch(response.status){
      case 200:
        return response.data.clients
      case 400:
        return "Credencial Invalida"
      default:

        return "Erro Interno"

    }
  }catch(error){
    throw(error)
  }
}


export const addClient = async (name:string,address:{[key:string]:string|number}, document: {[key:string]:string}) =>{
  try{
    const api = await createApiWithAuth()
    const data = {"name":name,"address":address,"document":document}
    const response = await api.post('/clients/add-new',data)
    
    switch (response.status){
      case 200:
        return true
      case 400:
        return "Credencial Invalida"
      case 409:
        return "Cliente ja Existe"
      default:
        return "Erro Interno"
    }
  }catch(error){
    throw error
  }
}


export const getEspecificClient = async(id:string) =>{
  try{

    const api = await createApiWithAuth()
    const data = {'id':id}
    
    const response = await api.post('/clients/get-especific-datas',data)

    switch(response.status){
      case 200:
        return response.data.clientInfos
      case 400:
        return "Credenciais Invalidas"
      case 404:
        return "Cliente Não Existe"
      default :
        return "Erro Interno"
    }
  }catch(error){
    throw error
  }
}

export const deleteClient = async(id:string,type:string) =>{
  try{
    const api = await createApiWithAuth()
    const data = {'id':id,'type':type}

    const response = await api.post('/clients/delete',data)

    switch(response.status){
      case 200:
        return "Cliente Excluido"
      case 401:
        return "Credenciais Invalidas"
      case 404:
        return "Erro Desconhecido"
      default:
        return "Erro Interno"
    }
  }catch(error){
    throw error
  }
}

File: Flask-app/routes/Clients.py
from flask import Flask, request, jsonify, Blueprint
from flask_cors import CORS
from cryptography.fernet import Fernet
from dotenv import load_dotenv
from jwt.exceptions import InvalidSignatureError
import os
from controllers.client_controller import ClientController
from controllers.cripto_controller import CriptographyController
load_dotenv()

clients_bp = Blueprint('clients',__name__)


def DescriptoToken(token):
    try:
        if not token:
            return False
        
        datas = CriptographyController().DecriptoDatas(token)
        if not datas:
            return False
    
        return datas
    except Exception as e:
        return False

@clients_bp.route('/get-all',methods=["POST"])
def GetClients():
    try:
        token = request.headers.get('Authorization')
        datas  = DescriptoToken(token)
        if not datas:
            return "Credenciais Invalidas", 401
        
        status, clients = ClientController().GetClients(datas['company_email'])
        if status:
            
            return jsonify({'status':'ok','clients':clients}),200
        return jsonify({'status':'error'}),500
    
    except InvalidSignatureError as i:
        return jsonify({'status': 'invalid'}), 400
        
    except Exception as e: 
        print('Error: ', e)
        return jsonify({'status':'error'}),500

@clients_bp.route('/add-new',methods=['POST'])
def AddNewClient():
    try:
        token = request.headers.get('Authorization')
        datas  = DescriptoToken(token)
        if not datas:
            return "Credenciais Invalidas", 401
        

        clients_datas = request.get_json()  
        created_client  = ClientController().AddNewClient(datas['company_email'],clients_datas['name'],clients_datas['address'],
                                                                clients_datas['document'])
        
        if created_client:
            return jsonify({'status':'ok'}),200
        else:
            return jsonify({'status':'ok'}),409
    
    except InvalidSignatureError as i:
        return jsonify({'status': 'invalid'}), 400
        
    except Exception as e:
        print('Error: ',e)
        return jsonify({'status':'error'}),500

@clients_bp.route('/get-especific-datas',methods=['POST'])
def GetEspecificDataClients():
    try:
        token = request.headers.get('Authorization')
        datas  = DescriptoToken(token)
        if not datas:
            return "Credenciais Invalidas", 401
        
        clients_datas = request.get_json()['id']
        tipo, client_id = clients_datas.split('&')
        clients_infos = ClientController().GetEspecicDataFromClient(datas['company_email'],client_id,tipo) 
        
        if clients_infos:
            return({'status':'ok','clientInfos':clients_infos}),200
        return jsonify({'status':'none'}),404
    except InvalidSignatureError as i:
        return jsonify({'status': 'invalid'}), 400
        
    except Exception as e:
        print('Error: ',e)
        return jsonify({'status':'error'}),500

@clients_bp.route('/delete',methods=['POST'])
def DeleteClient():
    try:
        token = request.headers.get('Authorization')
        datas = DescriptoToken(token)
        if not datas:
            return "Credenciais Invalidas" , 401
        
        client_id = request.get_json()['id']
        client_type = request.get_json()['type']    

        client_deleted = ClientController().DeleteClient(datas['company_email'],client_id,client_type)

        if client_deleted:
            return 'Cliente Excluido', 200
        else:
            return 'Erro Desconhecido',404
        
    except Exception as e:
        print('Error: ',e)
        return 'Erro Interno',500